cms/server/contest/handlers/contest.py

```diff
@@ -0,0 +1,386 @@
+from cms.server import compute_actual_phase, file_handler_gen
+from cms.locale import filter_language_codes
+from cmscommon.datetime import get_timezone, make_datetime, make_timestamp
+
+from .base import BaseHandler
+
+logger = logging.getLogger(__name__)
+
+
+NOTIFICATION_ERROR = "error"
+NOTIFICATION_WARNING = "warning"
+NOTIFICATION_SUCCESS = "success"
+
+
+def check_ip(ip, whitelist):
+    """Return if client IP belongs to one of the accepted subnets.
+
+    ip (string): IP address to verify.
+    whitelist (string): IP addresses or subnets to check against (separated
+        by a comma).
+
+    return (bool): whether client is equal to one of the IPs in the whitelist
+        or client belongs to one of the subnets in the whitelist.
+
+    """
+    for wanted in whitelist.split(","):
+        wanted, sep, subnet = wanted.partition('/')
+
+        subnet = 32 if sep == "" else int(subnet)
+        snmask = 2 ** 32 - 2 ** (32 - subnet)
+        wanted = struct.unpack(">I", socket.inet_aton(wanted))[0]
+        client = struct.unpack(">I", socket.inet_aton(ip))[0]
+
+        if (wanted & snmask) == (client & snmask):
+            return True
+
+    return False
+
+
+class ContestHandler(BaseHandler):
+    """A handler that has a contest attached.
+
+    Most of the RequestHandler classes in this application will be a
+    child of this class.
+    """
+    def prepare(self):
+        super(ContestHandler, self).prepare()
+
+        self.choose_contest()
+
+    def choose_contest(self, contest_name=None):
+        if self.application.service.contest is None:
+            if contest_name is None:
+                # Choose the contest found in the path argument
+                # see: https://github.com/tornadoweb/tornado/issues/1673
+                contest_name = self.path_args[0]
+
+            # Select the correct contest or return an error
+            try:
+                self.contest = self.contest_list[contest_name]
+            except KeyError:
+                # The right thing here would be:
+                #    raise tornado.web.HTTPError(404)
+                # however, that would make the "error.html" page fail because
+                # there is no self.contest available
+
+                # So, let's return to the contest list
+                self.redirect("/")
+                return
+        else:
+            # Select the contest specified on the command line
+            self.contest = Contest.get_from_id(self.application.service.contest,
+                                               self.sql_session)
+
+        # Run render_params() now, not at the beginning of the request,
+        # because we need contest_name
+        self.r_params = self.render_params()
+
+    def render_params(self):
+        ret = super(ContestHandler, self).render_params()
+
+        ret["contest"] = self.contest
+        ret["contest_root"] = ret["url_root"]
+        ret["real_contest_root"] = "/"
```

Comment:
Can you explain the semantic of `contest_root` and `real_contest_root`?
In some scenarios, this PR causes trouble with `StartHandler` and `LogoutHandler`.